### ISSUE TITLE
ICS6: Issues related to Solo Machine

### ISSUE THREAD

**Issue by josef-widder** [cosmos/ibc#524]
Surfaced from Informal Systems IBC Audit of cosmos-sdk hash https://github.com/cosmos/cosmos-sdk/commit/fe8a891f11dcefe708e43d53549beec808f98687.

## Specification

- [ ] there is a [spec](https://github.com/cosmos/cosmos-sdk/tree/master/x/ibc/light-clients/06-solomachine/spec) in the code repository. It has quite insightful preconditions and postconditions for the functions. These could be used to improve the accessibility of ICS006

- [ ] The [Verify](https://github.com/cosmos/ics/tree/master/spec/ics-006-solo-machine-client#state-verification-functions) functions have side effects (`clientState.consensusState.sequence++`). This is not the case in the verify functions of ICS002.
  
## Specification and code

- [ ] the naming in the [spec](https://github.com/cosmos/ics/tree/master/spec/ics-006-solo-machine-client#validity-predicate) does not match the naming in the [implementation](https://github.com/cosmos/cosmos-sdk/blob/fe8a891f11dcefe708e43d53549beec808f98687/x/ibc/light-clients/06-solomachine/types/update.go#L17).
     - the specification of this function talks about `assert`, which is understand as generating an error, while the code aborts the transactions which in the specification typically is written as `abortTransactionUnless`.

- [ ] governance [proposal](https://github.com/cosmos/cosmos-sdk/blob/fe8a891f11dcefe708e43d53549beec808f98687/x/ibc/light-clients/06-solomachine/types/proposal_handle.go) not discussed in the specification

**Comment by devashishdxt** [comment]
> This is true, and intentional. Do you think this is a bug, or unclearly documented, or something?

@cwgoes any comments on this: https://github.com/cosmos/cosmos-sdk/issues/9104

**Comment by devashishdxt** [comment]
> This is true, and intentional.

Also, may I ask the reason for this?

**Comment by ethanfrey** [comment]
> >   This is true, and intentional.
>
> Also, may I ask the reason for this?

This is a point for the spec in general - there is no explanation of the reasoning. It show the design, but it is very hard to follow *why* things do what they do or *how* to use them properly. It also makes it hard to distinguish bugs (undesired behavior) for intentional choices (thus this issue).

A document specifying *how* IBC should work on a high level would be very useful to follow the spec, or even use it as a reference when coding.

**Comment by colin-axner** [comment]
> This is a point for the spec in general - there is no explanation of the reasoning. It show the design, but it is very hard to follow _why_ things do what they do or _how_ to use them properly. It also makes it hard to distinguish bugs (undesired behavior) for intentional choices (thus this issue).
> 
> A document specifying _how_ IBC should work on a high level would be very useful to follow the spec, or even use it as a reference when architecting.

I believe the maintainers of this repository are very aware of the difficulty in reading the specifications and have begun discussion in how to supplement it with annotations or other documentation. I'm not sure if there is an issue to follow this work, but it is apart of the larger refactor of the spec repository

**Comment by colin-axner** [comment]
> @cwgoes any comments on this: https://github.com/cosmos/cosmos-sdk/issues/9104

The issue linked above is a bug in the implementation (the verify function side effects were not being referenced within a single handshake function). The spec is still correct, but documentation on side effects of the verify functions need to be documented clearly
